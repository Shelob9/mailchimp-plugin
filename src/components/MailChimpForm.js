import {CalderaForm} from "@calderajs/forms";
import React, {useState} from "react";
import {PacmanLoader} from 'react-spinners';
import PropTypes from 'prop-types';
/**
 * Component for stand-alone mailchimp forms served via Caldera API
 *
 * @param form
 * @param onChange
 * @param onBlur
 * @param onSubmit
 * @return {*}
 * @constructor
 */
function MailChimpForm({form, onChange, onBlur,onSubmit,hideOnSubmit}) {
	const [isSubmitting, setIsSubmitting] = useState(false);
	const [completed, setIsCompleted] = useState(false);
	const [message, setMessage] = useState('');

	const Spinner = () => (
		<div><PacmanLoader/></div>
	);

	if (!form.hasOwnProperty('fields')) {
		return <Spinner/>
	}

	const {processors} = form;
	const processor = processors.find(p => 'mc-subscribe' === p.type);

	if (completed) {
		return <div className={'success'}>{message}</div>
	}
	if (message) {
		return <div className={'error'}>{message}</div>
	}
	if (isSubmitting) {
		return <Spinner/>;
	}

	return (
		<div>
			<CalderaForm
				form={form}
				onSubmit={(
					//current values of all fields
					values,
					actions
				) => {
					setIsSubmitting(true);
					onSubmit(values,processor).then(r => r.json()).then(r => {
							if(hideOnSubmit){
								setMessage(r.message);
								setIsCompleted(r);
							}

						})
						.catch(e => {
							setIsSubmitting(false);
							if (e.hasOwnProperty('message')) {
								setMessage(e.message);
							} else {
								setMessage('An Error happened.');

							}
						});
				}}
				onChange={(values) => {
					console.log(values) //all field values
				}}
			/>
		</div>

	)
}

/**
 * Default Submit handler
 *
 * @param values
 * @return {*}
 */
const onSubmit = (values,processor) => {
	const getValue = (key) => {
		if (values.hasOwnProperty(key)) {
			return values[key];
		}
		return null;
	};
	const {listId, submitUrl} = processor;
	const mergeFields = {};
	const groupFields = {};
	processor.mergeFields.forEach(field => {
		mergeFields[field] = getValue(field);
	});
	processor.groupFields.forEach(field => {
		groupFields[field] = getValue(field);
	});


	const data = {
		email: getValue(processor.emailField),
		mergeFields,
		groupFields,
		listId
	};
	return fetch(submitUrl, {
		method: 'POST',
		headers: {
			'Content-Type': 'application/json',
		},
		body: JSON.stringify(data)
	});
};

MailChimpForm.propTypes = {
	form: PropTypes.object,
	onSubmit: PropTypes.func,
	onChange: PropTypes.func,
	onBlur: PropTypes.func,
};

MailChimpForm.defaultProps = {
	onSubmit,
	hideOnSubmit: true,
};
export default MailChimpForm;
