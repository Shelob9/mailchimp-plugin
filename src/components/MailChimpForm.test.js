import React from 'react';

import renderer from 'react-test-renderer';
import MailChimpForm from "./MailChimpForm";

import {mailChimpTestForm} from "./mailChimpTestForm.fixture";


describe( 'MailChimp mailChimpTestForm', () => {
	let onSubmit, onBlur,onChange;
	beforeEach( () => {
		onSubmit = jest.fn();
		onBlur = jest.fn();
		onChange = jest.fn();
	});

	it('Shows spinner if mailChimpTestForm does not have fields', () => {

		expect(renderer.create(
			<MailChimpForm
				form={{
					ID: 'mc-test-1'
				}}
				onBlur={onBlur}
				onChange={onChange}
				onSubmit={onSubmit}
			/>
		)).toMatchSnapshot();


	});

	it('Loads if it has proper mailChimpTestForm', () => {

		expect(renderer.create(
			<MailChimpForm
				form={mailChimpTestForm}
				onBlur={onBlur}
				onChange={onChange}
				onSubmit={onSubmit}
			/>
		)).toMatchSnapshot();


	});
});



