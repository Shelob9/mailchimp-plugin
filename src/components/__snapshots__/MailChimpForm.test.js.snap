// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MailChimp mailChimpTestForm Loads if it has proper mailChimpTestForm 1`] = `
<div>
  
  <div
    className="caldera"
  >
    <form
      className="mc-45907f0c59 caldera-form"
      onSubmit={[Function]}
    >
      <div
        className="sc-bdVaJa caldera-form-row caldera-row sc-bwzfXH pjOvZ"
      >
        <div
          className="caldera-column width-1 sc-bdVaJa iHZvIS"
          id="mc-45907f0c59-1-1"
        >
          <div
            className="caldera-field-wrapper caldera-field-wrapper-text"
          >
            <label
              className="caldera-field-label caldera-field-label-text"
              htmlFor="mc-email"
            >
              Email
               
            </label>
            <input
              className="caldera-field caldera-field-text"
              id="mc-email"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
            />
          </div>
        </div>
        <div
          className="caldera-column width-1 sc-bdVaJa iHZvIS"
          id="mc-45907f0c59-1-2"
        >
          <div
            className="caldera-field-wrapper caldera-field-wrapper-text"
          >
            <label
              className="caldera-field-label caldera-field-label-text"
              htmlFor="FNAME"
            >
              First Name
               
            </label>
            <input
              className="caldera-field caldera-field-text"
              id="FNAME"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
            />
          </div>
        </div>
      </div>
      <div
        className="sc-bdVaJa caldera-form-row caldera-row sc-bwzfXH pjOvZ"
      >
        <div
          className="caldera-column width-1 sc-bdVaJa iHZvIS"
          id="mc-45907f0c59-2-1"
        >
          <div
            className="caldera-field-wrapper caldera-field-wrapper-text"
          >
            <label
              className="caldera-field-label caldera-field-label-text"
              htmlFor="LNAME"
            >
              Last Name
               
            </label>
            <input
              className="caldera-field caldera-field-text"
              id="LNAME"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
            />
          </div>
        </div>
        <div
          className="caldera-column width-1 sc-bdVaJa iHZvIS"
          id="mc-45907f0c59-2-2"
        >
          <div
            className="caldera-field-wrapper caldera-field-wrapper-text"
          >
            <label
              className="caldera-field-label caldera-field-label-text"
              htmlFor="MMERGE3"
            >
              Primary Website
               
            </label>
            <input
              className="caldera-field caldera-field-text"
              id="MMERGE3"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
            />
          </div>
        </div>
      </div>
      <div
        className="sc-bdVaJa caldera-form-row caldera-row sc-bwzfXH pjOvZ"
      >
        <div
          className="caldera-column width-1 sc-bdVaJa iHZvIS"
          id="mc-45907f0c59-3-1"
        >
          <fieldset
            className="caldera-fieldset caldera-fieldset-checkboxes"
          >
            <legend>
              What React Framework Do You Want To Learn?
            </legend>
            <div
              className="caldera-field-group caldera-field-area-checkboxes caldera-field-wrapper caldera-field-wrapper-checkbox"
            >
              <input
                checked={false}
                className="caldera-field caldera-field-checkbox"
                id="75091baaad"
                onChange={[Function]}
                type="checkbox"
                value="1"
              />
              <label
                className="caldera-field-label caldera-field-label-checkbox"
                htmlFor="75091baaad"
              >
                Gatsby
                 
              </label>
            </div>
            <div
              className="caldera-field-group caldera-field-area-checkboxes caldera-field-wrapper caldera-field-wrapper-checkbox"
            >
              <input
                checked={false}
                className="caldera-field caldera-field-checkbox"
                id="0a4dc88500"
                onChange={[Function]}
                type="checkbox"
                value="1"
              />
              <label
                className="caldera-field-label caldera-field-label-checkbox"
                htmlFor="0a4dc88500"
              >
                Next.Js
                 
              </label>
            </div>
            <div
              className="caldera-field-group caldera-field-area-checkboxes caldera-field-wrapper caldera-field-wrapper-checkbox"
            >
              <input
                checked={false}
                className="caldera-field caldera-field-checkbox"
                id="5a44f196ec"
                onChange={[Function]}
                type="checkbox"
                value="1"
              />
              <label
                className="caldera-field-label caldera-field-label-checkbox"
                htmlFor="5a44f196ec"
              >
                Create React App
                 
              </label>
            </div>
          </fieldset>
        </div>
      </div>
      <div
        className="sc-bdVaJa caldera-form-row caldera-row sc-bwzfXH pjOvZ"
      >
        <div
          className="caldera-column width-1 sc-bdVaJa iHZvIS"
          id="mc-45907f0c59-4-2"
        >
          <div
            className="caldera-field-group caldera-field-area-submit caldera-field-wrapper caldera-field-wrapper-submit"
          >
            <label
              className="caldera-field-label caldera-field-label-submit"
              htmlFor="mc-submit"
            >
              
               
            </label>
            <input
              className="caldera-field caldera-field-text"
              disabled={false}
              id="mc-submit"
              type="submit"
            />
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;

exports[`MailChimp mailChimpTestForm Shows spinner if mailChimpTestForm does not have fields 1`] = `
<div>
  <div
    className="css-16tm1g7-wrapper-wrapper"
  >
    <div
      className="css-1anlcob-pacmanStyle"
    />
    <div
      className="css-1oj2eaa-pacmanStyle"
    />
    <div
      className="css-1y2ttxx-ballStyle"
    />
    <div
      className="css-im5qma-ballStyle"
    />
    <div
      className="css-10tjdos-ballStyle"
    />
    <div
      className="css-x3nw5t-ballStyle"
    />
  </div>
</div>
`;
